### ISSUE TITLE
Parallel tree_map

### ISSUE THREAD

**Issue by JGameCreation** [google/jax#11394]
`tree_map` is a very important and frequently used jax function. Although we know that all the leaves will be processed independently, it is just implemented as a simple for loop right now. Would  it be possible to evaluate the map function on multiple leaves in parallel?
I don't know if the jit-compiler already does some auto-parallelization magic but from the runtimes it doesn't seem like it does and I also couldn't find any information on this in the documentation.

**Comment by YouJiacheng** [comment]
`tree_map` isn't specified with parallel or sequential semantic.
Currently it is backed by list comprehension, thus its python execution is sequential.
However, JIT compiler may execute code without data dependency in parallel.
And JAX has asynchronous dispatch for non-JIT(op-by-op) execution, which allows parallel execution.

**Comment by hawkinsp** [comment]
I think it's unlikely we will add our own parallel tree map. And as @YouJiacheng suggests, inside something like a `jit` there would be no benefit.

However, if you wanted to have parallelism outside a `jit`, you could build such a function using `tree_flatten`, a parallel map of your choice over, say, lists, and `tree_unflatten`. It should only be a few lines of code.

Hope that helps!

**Comment by DiagRisker** [comment]
I believe this  
> JAX has asynchronous dispatch for non-JIT(op-by-op) execution, which allows parallel execution

 should be in the documentation. At the current moment I don't find it.

Additionally, as it's not a guaranteed parallel execution per say (if yes please documentation will help), it would really be a big improvement for JAX to have a map method which can parallelize without differentiation (vmap and pmap, jax.jit do), nor uniform explicit shape requirements . 
AT the moment, I am forced to use python standard library for my applications (when I can't deal with `jax.Array` type) :
For instance, I am using massively JAX for non ML code such as rendering (pixel wise), and this would come handy (compilation of my code will include shape data, and or rng keys (Non-hashable static arguments are not supported.) ).

**Comment by jakevdp** [comment]
> I believe this should be in the documentation. At the current moment I don't find it.

Asynchronous dispatch is mentioned in the documentation here: https://docs.jax.dev/en/latest/async_dispatch.html and referenced inmany places

> it would really be a big improvement for JAX to have a map method which can parallelize without differentiation (vmap and pmap, jax.jit do)

I don't understand what you mean by "parallelize without differentiation"

> and this would come handy (compilation of my code will include shape data, and or rng keys (Non-hashable static arguments are not supported.) ).

I would suggest opening a [new issue](https://github.com/jax-ml/jax/issues/new/choose) with examples of the code you're running and the problems you're having with it. It's hard to understand the full situation from this short description.

**Comment by DiagRisker** [comment]
> Asynchronous dispatch is mentioned in the documentation here: https://docs.jax.dev/en/latest/async_dispatch.html and referenced in many places

You are right, the info asynchronous dispatch info is there, however it was not explicitly clear to me, that jax.jit will parallelize "map" with it (my bad in a way).

> I don't understand what you mean by "parallelize without differentiation

A small example, is better, I'll do a full new issue, because it is indeed not just about the operation being differentiable with respect to the input.
